### ISSUE TITLE
Composition of graphs

### ISSUE THREAD

**Issue by vporton** [jgrapht/jgrapht#464]
I need composition of digraphs (I mean composition of binary relations represented as digraphs.) in my project. It seems that JGraphT is missing this operation.

I remind that composition is defined by the formula

x(R*S)z <=> exists y: (xSy and yRz)

**Comment by jkinable** [comment]
I presume this issue can be closed.

**Comment by vporton** [comment]
@jkinable Why have you closed it without implementing the requested feature?!

**Comment by jkinable** [comment]
My bad, it was referring to #465, which you closed.

Nevertheless, based on your very limited description, it is unclear what feature you are missing. Hence, I doubt that anyone will follow this up. So I suggest you provide a proper description of what you are looking for, preferably with a reference to a relevant article/webpage.

**Comment by vporton** [comment]
Let S: Graph<V,E1> and R: Graph<V,E2> be two graphs with edge types E1 and E2 and the same vertex type V. Let C (which can be called "composition of edges") be a function of the signature (E1,E2) -> E3.

I want a function (which I call "composition") which creates graph T: Graph<V,E3> whose edges are all C(e1,e2) such that the target of e1 is .equal() to the source of e2.

For the special case of E1=E2=E3=Void (and the evident function C which maps two Void values into Void value) and S and T being of DirectedGraph type, composition is the well known mathematical compostion of binary relations (DirectedGraph<V,Void> with Void edges is essentially the same as a finite binary relation on V).

There should be special version of composition for E1=E2=E3=Void and the evident C for this case, because this special case is most often used in mathematics.

**Comment by jbduncan** [comment]
Hi @vporton, I cannot speak for the JGraphT team, but for me it's not really clear what you're after.

I understand from your description that if given a graph _A_ like

![capture](https://user-images.githubusercontent.com/2482228/34590734-87ae1574-f1b0-11e7-9248-86d007ca0b7d.PNG)

a graph _B_ like

![capture2](https://user-images.githubusercontent.com/2482228/34590739-8c469a98-f1b0-11e7-9fdd-96e53ccd1ba6.PNG)

and an edge composition function like

```java
public static String composeEdges(int e1, char e2) {
  // using Guava's com.google.common.base.Strings#repeat
  return Strings.repeat(Character.toString(e2), e1);
}
```

then a new graph of some kind should be created from those constituent parts, but I'm not clear on what the graph should look like, and how it would use graphs _A_ and _B_ and the composition function.

Could you clarify things for me?

**Comment by vporton** [comment]
The composition of your two graphs is an empty graph, because no destination of an edge of A is a source of an edge of B (the sets of vertices of A and B do not overlap).

Well, if the JGraphT team won't do my request, I am OK to create a GitHub fork myself. But it is better to be done by more experienced in this people.

**Comment by jbduncan** [comment]
Oh cool, thanks for the clarification!

So if I tweaked _A_ and _B_ so they have the same node and edge types, perhaps something like this:

**A**
![capture](https://user-images.githubusercontent.com/2482228/34591219-ec1851ac-f1b3-11e7-9180-63eecad5db88.PNG)

**B**
![capture2](https://user-images.githubusercontent.com/2482228/34591222-f0636788-f1b3-11e7-925f-8784108902eb.PNG)

...and if I changed the composition function to:

```java
public static String composeEdges(char e1, char e2) {
  return new StringBuilder(2).append(e1).append(e2).toString();
}
```

...what would the result look like? :)

**Comment by vporton** [comment]
@jbduncan 

![graph](https://user-images.githubusercontent.com/2900574/34591521-df98e186-f1c6-11e7-9b81-d742c8912b06.png)

**Comment by jbduncan** [comment]
@vporton Great, that clears things up a bit for me. Many thanks! :+1:

Are there any edge cases here that I'm missing (other than the case where if the two input graphs' vertices don't overlap, it then produces an empty graph)?

**Comment by vporton** [comment]
@jbduncan As far as I understand, there are no "edge cases" (whatever this may mean).

However, if there are multiple edges between given edges, the number of edges multiplies.

**Comment by jbduncan** [comment]
Hrm, okay. I suppose what I really wanted to ask is not what "edge cases" there may be, but rather what the general algorithm is for composing graphs like this in layman's terms, because although I did study graphs at university, it's been such a long time for me since I studied any relevant theory that I'm struggling to understand your [earlier mathematical description](https://github.com/jgrapht/jgrapht/issues/464#issuecomment-355436565).

Is this something that you can describe for me?

**Comment by vporton** [comment]
Binary relation G is monovalued iff there are no x and two different p and q such that both xGp and xGq.

Composition (see my original issue) of monovalued binary relations coincides with the well known composition of (monovalued) functions: (R*S)(x)=R(S(x)).

Composition of binary relations is a generalization of composition of functions.

See also https://en.wikipedia.org/wiki/Composition_of_relations

**Comment by jbduncan** [comment]
Okay, so am I right to think that if we put the edge-composition function to one side for now, the algorithm expressed as Java-like pseudocode using JGraphT would look something like this?

```java
Graph<Integer, DefaultEdge> R = graph constructed from edges {(1,1),(1,4),(2,3),(3,1),(3,4)}
Graph<Integer, DefaultEdge> S = graph constructed from edges {(1,0),(2,0),(3,1),(3,2),(4,1)}
Graph<Integer, DefaultEdge> composition = empty graph
for (var edgeR : R.edgeSet()) {
  for (var edgeS : S.edgeSet()) {
    if (R.getEdgeTarget(edgeR).equals(S.getEdgeSource(edgeS)) {
      composition.addEdge(R.getEdgeSource(edgeR), S.getEdgeTarget(edgeS);
    }
  }
}

println(composition); // -> {(1,0),(1,1),(2,1),(2,2),(3,0),(3,1)}
```

**Comment by vporton** [comment]
Yes, this is graph composition

**Comment by d-michail** [comment]
Hi, 

Not sure whether someone has the time to implement this right now. I would suggest to create a prototype implementation in your own fork and then create a PR (pull-request). Some directions: 
  - create a class Composition in package 'org.jgrapht.alg'
  - do not use the singleton pattern like `TransitiveClosure`, that class needs some refactoring which we had no time to perform yet
  - the class should have two constructors, one with an edge-composition function and one without. Based on the constructor call initialisations, a method should create the new graph
 - you can start with the simple nested-loops algorithm that @jbduncan described but you can also write something like [hash-join](https://en.wikipedia.org/wiki/Hash_join). The hash-join method would be easier to write for the case that you do not have an edge-composition function.
- the edge composition function should be of the `BiPredicate` interface so that it can be used with lambdas
- make sure you use generics properly
- write a test class in the same package which tests that the class works as expected with as least 2-3 different inputs as well as some "edge" cases

Hope this helps
